File: dotnet/src/MyDotey.NameCache/NameCache.cs
using System;
using System.Collections.Concurrent;
using System.Linq;
using System.Text;

namespace MyDotey.NameCache
{
    public class NameCache
    {
        private NameCacheNode<object, CachedName> _namePartTree = new NameCacheNode<object, CachedName>();
        private string _separator;

        public NameCache()
            : this(null)
        {

        }

        public NameCache(string separator)
        {
            _separator = separator == null ? String.Empty : separator.Trim();
        }

        public CachedName Get(params object[] nameParts)
        {
            if (nameParts == null || nameParts.Length == 0)
                throw new ArgumentException("nameParts is null or empty");

            NameCacheNode<Object, CachedName> node = _namePartTree;
            for (int i = 0; i < nameParts.Length; i++)
            {
                node = node.Children.GetOrAdd(nameParts[i], n => new NameCacheNode<Object, CachedName>(n));
            }

            if (node.Value == null)
            {
                lock (node)
                {
                    if (node.Value == null)
                        node.Value = new CachedName(_separator, nameParts);
                }
            }

            return node.Value;
        }
    }

}

File: java/name-cache/src/main/java/org/mydotey/namecache/NameCache.java
package org.mydotey.namecache;

import java.util.Objects;
import java.util.function.Function;

/**
 * @author koqizhao
 *
 * Jul 20, 2018
 */
public class NameCache {

    private NameCacheNode<Object, CachedName> _namePartTree = new NameCacheNode<Object, CachedName>();
    private String _separator;

    private Function<Object, NameCacheNode<Object, CachedName>> _nodeCreator = k -> new NameCacheNode<Object, CachedName>(
            k);

    public NameCache() {
        this(null);
    }

    public NameCache(String separator) {
        _separator = separator == null ? "" : separator.trim();
    }

    public CachedName get(Object... nameParts) {
        if (nameParts == null || nameParts.length == 0)
            Objects.requireNonNull("nameParts is null or empty");

        NameCacheNode<Object, CachedName> node = _namePartTree;
        for (int i = 0; i < nameParts.length; i++) {
            node = node.getChildren().computeIfAbsent(nameParts[i], _nodeCreator);
        }

        if (node.getValue() == null) {
            synchronized (node) {
                if (node.getValue() == null)
                    node.setValue(new CachedName(_separator, nameParts));
            }
        }

        return node.getValue();
    }

}
